{
    "active_lock_reason": null,
    "assignee": null,
    "assignees": [],
    "author_association": "MEMBER",
    "body": "### Describe the bug\n\nIt's possible to get an infinite recursion error on binops https://github.com/pandas-dev/pandas/pull/56082 that doesn't exist in Python\n\n### Code to reproduce the behaviour:\n\n```cython\r\nimport cython\r\n\r\n@cython.cclass\r\nclass C:\r\n    def __add__(self, other):\r\n        return NotImplemented\r\n\r\n    def __radd__(self, other):\r\n        return self.__add__(other)\r\n\r\nC() + C()\r\n```\r\n\n\n### Expected behaviour\n\nIn Python this correctly gives\r\n\r\n```\r\nTypeError: unsupported operand type(s) for +: 'notimplemented.C' and 'notimplemented.C'\r\n```\r\n\r\nIn Cython it gives\r\n\r\n```\r\n  File \"notimplemented.pyx\", line 17, in notimplemented.C.__radd__\r\n    return self.__add__(other)\r\n  File \"notimplemented.pyx\", line 17, in notimplemented.C.__radd__\r\n    return self.__add__(other)\r\n  File \"notimplemented.pyx\", line 17, in notimplemented.C.__radd__\r\n    return self.__add__(other)\r\n  [Previous line repeated 997 more times]\r\nRecursionError: maximum recursion depth exceeded while calling a Python object\r\n```\r\n\r\nI think the issue is:\r\n\r\n```\r\nstatic PyMethodDef __pyx_methods_14notimplemented_C[] = {\r\n  {\"__radd__\", (PyCFunction)__pyx_pw_14notimplemented_1C_3__radd__, METH_O|METH_COEXIST, 0},\r\n  {\"__reduce_cython__\", (PyCFunction)(void*)(__Pyx_PyCFunction_FastCallWithKeywords)__pyx_pw_14notimplemented_1C_5__reduce_cython__, __Pyx_METH_FASTCALL|METH_KEYWORDS, 0},\r\n  {\"__setstate_cython__\", (PyCFunction)(void*)(__Pyx_PyCFunction_FastCallWithKeywords)__pyx_pw_14notimplemented_1C_7__setstate_cython__, __Pyx_METH_FASTCALL|METH_KEYWORDS, 0},\r\n  {0, 0, 0, 0}\r\n};\r\n```\r\n\r\nHere we need to add `__add__` and make it point to the `__add__` function rather than our binop wrapper function. This will mean manually calling `C.__add__` skips the binop wrapper and calls the function exactly as written.\n\n### OS\n\nLinux\n\n### Python version\n\n3.11.5\n\n### Cython version\n\nMost recent master (but I think also 3.0.x)\n\n### Additional context\n\n_No response_",
    "closed_at": null,
    "comment_data": [],
    "comments": 0,
    "comments_url": "https://api.github.com/repos/cython/cython/issues/5863/comments",
    "created_at": "2023-11-23T18:54:55Z",
    "event_data": [],
    "events_url": "https://api.github.com/repos/cython/cython/issues/5863/events",
    "html_url": "https://github.com/cython/cython/issues/5863",
    "id": 2008689453,
    "labels": [],
    "labels_url": "https://api.github.com/repos/cython/cython/issues/5863/labels{/name}",
    "locked": false,
    "milestone": null,
    "node_id": "I_kwDOABDGAc53uist",
    "number": 5863,
    "performed_via_github_app": null,
    "reactions": {
        "+1": 0,
        "-1": 0,
        "confused": 0,
        "eyes": 0,
        "heart": 0,
        "hooray": 0,
        "laugh": 0,
        "rocket": 0,
        "total_count": 0,
        "url": "https://api.github.com/repos/cython/cython/issues/5863/reactions"
    },
    "repository_url": "https://api.github.com/repos/cython/cython",
    "state": "open",
    "state_reason": null,
    "timeline_url": "https://api.github.com/repos/cython/cython/issues/5863/timeline",
    "title": "[BUG] Infinite recursion in binops",
    "updated_at": "2023-11-23T18:54:55Z",
    "url": "https://api.github.com/repos/cython/cython/issues/5863",
    "user": {
        "avatar_url": "https://avatars.githubusercontent.com/u/10536947?v=4",
        "events_url": "https://api.github.com/users/da-woods/events{/privacy}",
        "followers_url": "https://api.github.com/users/da-woods/followers",
        "following_url": "https://api.github.com/users/da-woods/following{/other_user}",
        "gists_url": "https://api.github.com/users/da-woods/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/da-woods",
        "id": 10536947,
        "login": "da-woods",
        "node_id": "MDQ6VXNlcjEwNTM2OTQ3",
        "organizations_url": "https://api.github.com/users/da-woods/orgs",
        "received_events_url": "https://api.github.com/users/da-woods/received_events",
        "repos_url": "https://api.github.com/users/da-woods/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/da-woods/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/da-woods/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/da-woods"
    }
}